src/shared/bus-unit-util.c

```diff
@@ -2937,3 +2938,73 @@ int bus_service_manager_reload(sd_bus *bus) {
 
         return 0;
 }
+
+int unit_freezer_restore(const char *name, UnitFreezer *ret) {
+        _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
+        _cleanup_free_ char *namedup = NULL;
+        int r;
+
+        assert(name);
+        assert(ret);
+        assert(!ret->name);
+        assert(!ret->bus);
+
+        namedup = strdup(name);
+        if (!namedup)
+                return log_oom_debug();
+
+        r = bus_connect_system_systemd(&bus);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open connection to systemd: %m");
+
+        (void) sd_bus_set_method_call_timeout(bus, FREEZE_TIMEOUT);
+
+        *ret = (UnitFreezer) {
+                .name = TAKE_PTR(namedup),
+                .bus = TAKE_PTR(bus),
+        };
+        return 0;
+}
+
+void unit_freezer_cancel(UnitFreezer *f) {
+        if (!f)
+                return;
+
+        f->name = mfree(f->name);
+        f->bus = sd_bus_flush_close_unref(f->bus);
+}
+
+int unit_freezer_freeze(const char *name, UnitFreezer *ret) {
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(unit_freezer_cancel) UnitFreezer f = UNIT_FREEZER_NULL;
+        int r;
+
+        assert(ret);
+
+        r = unit_freezer_restore(name, &f);
+        if (r < 0)
+                return r;
+
+        r = bus_call_method(f.bus, bus_systemd_mgr, "FreezeUnit", &error, NULL, "s", f.name);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to freeze unit %s: %s", f.name, bus_error_message(&error, r));
+
+        *ret = TAKE_STRUCT(f);
+        return 1;
+}
+
+int unit_freezer_thaw(UnitFreezer *_f) {
```

Comment:
So I think this function should be generic and not take ownership of the freezer. Instead please introduce unit_freezer_thaw_done that returns void and calls unit_freezer_done at the end.